net/aria2/files/aria2.init

```diff
@@ -1,112 +1,348 @@
-# Copyright (C) 2016 kuoruan <kuoruan@gmail.com>
+# Copyright (C) 2016-2017 Hsing-wang Liao <kuoruan@gmail.com>
+# Licensed to the public under the Apache License 2.0.
 
 START=99
-SERVICE_WRITE_PID=1
-SERVICE_DAEMONIZE=1
-
-
-append_params() {
-	local p; local v; local s="$1"; shift
-	for p in $*; do
-		config_get v "$s" "$p"
-		[ -n "$v" ] && (
-			p=$(echo "$p" | sed -e 's|_|-|g');
-			echo "$p=$v" >> $config_file
-			)
+USE_PROCD=1
+
+NAME=aria2
+PROG=/usr/bin/aria2c
+
+_info() {
+	logger -p daemon.info -t "$NAME" "$*"
+}
+
+_err() {
+	logger -p daemon.err -t "$NAME" "$*"
+}
+
+_make_dir() {
+	local d
+	for d in "$@"; do
+		if [ ! -d "$d" ]; then
+			mkdir -p "$d" 2>/dev/null || return 1
+		fi
 	done
+
+	return 0
 }
 
-section_enabled() {
-	local result
-	config_get_bool result "$1" 'enabled' 0
-	[ $result -eq 1 ]
+_create_file() {
+	touch "$@" 2>/dev/null
 }
 
-option_disabled() {
-	local result
-	config_get_bool result "$1" "$2" 1
-	[ $result -eq 0 ]
+_change_owner() {
+	local u="$1"; shift
+
+	local d
+	for d in "$@"; do
+		if [ -f "$d" ]; then
+			chown "$u" "$d" 2>/dev/null || return 1
+		elif [ -d "$d" ]; then
+			chown -R "$u" "$d" 2>/dev/null || return 1
+		fi
+	done
+
+	return 0
 }
 
-start_instance() {
+_change_file_mode() {
+	local mod="$1"; shift
+	chmod "$mod" "$@" 2>/dev/null
+}
+
+_reset_dir_mode() {
+	local d
+	for d in "$@"; do
+		if [ -d "$d" ]; then
+			find "$d" -type d -exec chmod 755 {} \; 2>/dev/null
+			find "$d" -type f -exec chmod 644 {} \; 2>/dev/null
+		fi
+	done
+}
+
+append_options() {
+	local o; local v
+	for o in "$@"; do
+		v="$(eval echo "\$$o")"
+		[ -n "$v" ] && \
+			echo "${o//_/-}=$v" >>"$config_file_tmp"
+	done
+}
+
+append_setting() {
 	local s="$1"
-	local user
+	[ -n "$s" ] && \
+		echo "$s" >>"$config_file_tmp"
+}
+
+append_header() {
+	local h="$1"
+	[ -n "$h" ] && \
+		echo "header=\"${h}\"" >>"$config_file_tmp"
+}
+
+aria2_validate() {
+	uci_validate_section "$NAME" aria2 "$1" \
+		'enabled:bool:0' \
+		'enable_logging:bool' \
+		'enable_proxy:bool' \
+		'config_dir:string:/var/etc/aria2' \
+		'user:string' \
+		'all_proxy:string' \
+		'all_proxy_passwd:string' \
+		'all_proxy_user:string' \
+		'auto_save_interval:range(0,600)' \
+		'bt_enable_lpd:or("true","false")' \
+		'bt_max_open_files:uinteger' \
+		'bt_max_peers:uinteger' \
+		'bt_remove_unselected_file:or("true","false")' \
+		'bt_request_peer_speed_limit:string' \
+		'bt_save_metadata:or("true","false")' \
+		'bt_seed_unverified:or("true","false")' \
+		'bt_stop_timeout:uinteger' \
+		'bt_tracker:list(string)' \
+		'ca_certificate:file' \
+		'certificate:file' \
+		'check_certificate:or("true","false"):true' \
+		'connect_timeout:uinteger' \
+		'dht_listen_port:string' \
+		'dir:string' \
+		'disable_ipv6:or("true","false")' \
+		'disk_cache:string' \
+		'enable_dht:or("true","false"):true' \
+		'enable_dht6:or("true","false")' \
+		'enable_peer_exchange:("true","false")' \
+		'event_poll:("epoll","kqueue","port","poll","select")' \
```

Comment:
Missing `or` in the above 2 datatype spec?

Reply:
@yousong OK. I will make a new PR.